# Promise

- 자바스크립트는 비동기 처리를 위한 하나의 패턴으로 콜백함수를 사용한다.

  > 전통적인 콜백 패턴은 가독성이 나쁘고 비동기 처리 중 발생한 에러의 예외 처리가 곤란하며 여러 개의 비동기 처리를 한번에 처리하는 것도 한계가 있다.

- ES6에서는 비동기처리를 위해 프로미스(Promise)를 도입했다.

  > 프로미스는 전통적인 콜백 패턴이 가진 단점을 보완하며 비동기 처리 시점을 명확하게 표현할 수 있는 장점이 있다.

</br>

## 비동기 처리를 위한 콜백 패턴의 단점

### 콜백 헬

---

- GET 요청을 위한 get 함수는 비동기 함수이다.

  비동기 함수 내에서 서버의 응답 결과를 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.

- 비동기 함수인 get이 취득한 서버로부터의 응답은, 응답에 대한 후속 처리를 위한 콜백함수를 비동기 함수인 get에 전달해야한다. 

  ``` javascript
  const get = (url, callback) => {
      const xhr = new XMLHttpRequest();
      xhr.open('GET', url);
      xhr.send();
      
      xhr.onload = () => {
          if (xhr.status === 200) {
              //서버의 응답을 콜백함수에 전달하면서 호출하여 응답에 대한 후속처리를 한다.
              callback(JSON.parse(xhr.response));
              } else {
                  console.error(`${xhr.status} ${xhr.statusText}`);
              }
      };
  };
  
  // id가 1인 post 취득
  // 서버의 응답에 대한 후속 처리를 위한 콜백함수를 비동기 함수인 get에 전달해야한다.
  get('https://jsonplaceholder.typicode.com/posts/1', console.log);
  ```

- 콜백 헬(callback hell) 이란?

  : 비동기 함수의 처리 결과를 가지고 다른 비동기 함수를 호출해야하는 경우,

  **함수 호출이 중첩(nesting)**이 되어 복잡도가 높아지는 현상이 발생한다.

- 콜백 헬은 가독성을 나쁘게 하며 실수를 유발하는 원인이 된다

  ``` javascript
  get('/step1', a => {
    get(`/step2/${a}`, b => {
      get(`/step3/${b}`, c => {
        get(`/step4/${c}`, d => {
          console.log(d);
        });
      });
    });
  });
  ```

</br>

### 에러 처리의 한계

---

- 비동기 처리를 위한 콜백 패턴은 에러 처리가 곤란하다.

- 에러는 호출자(caller)방향으로 전파된다.(콜 스택 아래 방향)

  setTimeout 함수는 비동기함수이며 콜백함수의 호출자가 아니므로

  콜백함수가 발생시킨 에러는 catch 블록에서 캐치되지않는다.

  ``` javascript
  try {
    setTimeout(() => { throw new Error('Error!'); }, 1000);
  } catch (e) {
    // 에러를 캐치하지 못한다
    console.error('캐치한 에러', e);
  }
  ```

</br></br>

## 프로미스의 생성

- 비동기 처리를 위한 콜백 패턴의 콜백 헬이나 에러 처리의 한계를 극복하기 위해 ES6에서 프로미스(Promise)가 도입되었다.

- 프로미스는 Promise 생성자 함수를 통해 인스턴스화 한다.

- Promise 생성자 함수는 비동기 처리를 수행할 콜백 함수를 인자로 전달받는데 이 콜백 함수는 resolve와 reject 함수를 인수로 전달받는다.

  ``` javascript
  // Promise 객체 생성
  const promise = new Promise((resolve, reject) => {
      // Promise 함수의 콜백 함수 내부에서 비동기 처리를 수행한다.
      if (/*비동기 처리 성공*/) {
        resolve('result');
       } else {
        /* 비동기 처리 실패*/
        reject('failure reason');
       }
  });
  ```

- Promise 생성자 함수가 인수로 전달받은 콜백 함수 내부에서 비동기 처리를 수행한다.

  - 비동기 처리가 성공하면 콜백 함수의 인수로 전달받은 resolve 함수를 호출
  -  비동기 처리가 실패하면 reject 함수를 호출한다. 

  ``` javascript
  // GET 요청을 위한 비동기 함수
  const promiseGet = url => {
      return new Promise((resolve, reject) => {
          const xhr = new XMLHttpRequest();
          xhr.open('GET', url);
          xhr.send();
          
          xhr.onload = () => {
            if (xhr.status === 200) {
                // 성공적으로 응답을 전달받으면 resolve 함수를 호출
                resolve(JSON.parse(xhr.response));
            } else {
                // 에러 처리를 위해 reject 함수를 호출한다.
                reject(new Error(xhr.status));
            }
          };
      });
  };
  
  // promiseGet는 Promise 객체를 반환한다.
  promiseGet('https://jsonplaceholder.typicode.com/posts/1');
  ```

- 프로미스로 구현된 비동기 함수는 함수 내부에서 Promise 객체를 생성하고 반환한다.

- 비동기 처리는 Promise 생성자 함수가 인수로 전달받은 콜백함수 내부에서 수행한다. 

- 프로미스로 구현된 비동기함수가 반환하는 Promise 객체는 

  비동기 처리가 성공(fulfilled)했는지, 실패(rejected)했는지 등의 상태(state)정보를 갖는다. 

  | 상태      | 의미                                                         | 구현                                               |
  | --------- | ------------------------------------------------------------ | -------------------------------------------------- |
  | pending   | 비동기 처리가 아직 수행되지 않은 상태                        | resolve 또는 reject 함수가 아직 호출되지 않은 상태 |
  | fulfilled | 비동기 처리가 수행된 상태(성공)                              | resolve 함수가 호출된 상태                         |
  | rejected  | 비동기 처리가 수행된 상태(실패)                              | reject 함수가 호출된 상태                          |
  | settled   | fulfilled 또는 rejected와 상관없이 pending이 아닌 상태, 즉 비동기 처리가 수행된 상태 | resolve 또는 reject 함수가 호출된 상태             |

- Promise 객체의 상태 정보는 resolve 또는 reject 함수를 호출하는 것으로 결정된다.
- resolve 또는 reject 함수를 호출할 때 전달한 비동기 처리 결과 또는 에러는 Promise 객체의 후속 처리 메서드에게 전달된다.

</br></br>

### 프로미스의 후속 처리 메서드

---

- 프로미스로 구현된 비동기 함수는 **Promise 객체를 반환**해야한다.

- 프로미스로 구현된 비동기 함수를 호출하는 측(promise consumer)에서는

  Promise 객체의 **후속 처리 메서드 then, catch, finally를 통해 비동기 처리 결과 또는 에러 메시지를 전달 받아 후속 처리를 수행**한다.

- Promise 객체는 비동기 처리가 성공(fulfilled)했는지 또는 실패(rejected)했는지 등의 상태 정보를 갖는다.

  이 **상태 정보에 따라 후속처리 메서드를 체이닝 방식으로 호출**한다.

</br>

**프로미스의 후속처리 메서드**

- Promise.prototype.then

  - then 메서드는 두 개의 콜백 함수를 인수로 전달 받는다.
    - 첫번째 콜백함수는 프로미스가 fulfilled 상태(resolve 함수가 호출된 상태)일 때 호출되고,
    - 두번째 콜백함수는 프로미스가 rejected 상태(rejected 함수가 호출된 상태)일 때 호출된다.

  - then 메서드는 언제나 Promise 객체를 반환한다.

  - then 메서드의 콜백함수가 Promise 객체가 아닌 값을 반환하면

    그 값을 resolve 또는 reject 하여 Promise 객체를 반환한다.

``` javascript
// fulfilled
new Promise(resolve => resolve('fulfilled'))
    .then(v => console.log(v), e => console.log(e)); // fulfilled

// rejected
new Promise((_, reject) => reject(new Error('rejected')))
    .then(v => console.log(v), e => console.log(e));
// Error: rejected
```

- Promise.prototype.catch

  - catch 메서드는 한 개의 콜백 함수를 인수로 전달 받는다.

  - catch 메서드의 콜백함수는 예외(비동기 처리에서 발생한 에러와 then 메서드에서 발생한 에러)가 발생하면 호출된다.

  - catch 메서드는 then(undefined, onRejected)과 동일하게 동작한다.

    then 메서드와 마찬가지로 언제나 Promise 객체를 반환한다.

  ``` javascript
  //rejected
  new Promise((_, reject) => reject(new Error('rejected')))
      .catch(e => console.log(e));
  // Error: rejected
  ```

- Promise.prototype.finally

  - finally 메서드는 한 개의 콜백 함수를 인수로 전달 받는다.

  - finally 메서드의 콜백 함수는 프로미스의 성공(fulfilled) 또는 실패(rejected)와 상관없이 무조건 한번 호출된다.

  - finally 메서드는 프로미스의 상태와 상관없이 공통적으로 수행해야할 처리를 할 때 유용하다.

  - finally 메서드도 then / catch 메서드와 마찬가지로 언제나 Promise 객체를 반환한다.

    ``` javascript
    new Promise(() => {})
        .finally(() => console.log('finally')); //finally
    ```

</br>

**프로미스로 구현한 비동기 함수 get을 사용한 후속처리**

``` javascript
const promiseGet = url => {
    return new Promise((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        xhr.open('GET', url);
        xhr.send();
        
        xhr.onload = () => {
            if (xhr.status === 200) {
              // 성공적으로 응답전달시 resolve 함수 호출
             resolve(JSON.parse(xhr.response));
        	} else {
             // 에러 처리를 위해 reject 함수호출
             reject(new Error(xhr.status));
            } 
        };
    });
};

// promiseGet는 Promise 객체를 반환한다.
promiseGet('https://jsonplaceholder.typicode.com/posts/1')
    .then(res => console.log(res))
	.catch(err => console.log(err))
	.finally(() => console.log('bye'));
```

</br></br>

## 프로미스의 에러처리

- (위의 프로미스로 구현한 비동기 함수 Get(promiseGet)이어서.. )

- 비동기 처리 시에 발생한 에러는 then 메서드의 두번째 콜백 함수로 처리할 수 있다.

  ``` javascript
  const wrongUrl = 'https://jsonplaceholder.typicode.com/XXX/1';
  
  // 부적절한 url이 지정되어 에러발생
  promiseGet(wrongUrl)
  	.then(res => console.log(res), err => console.log(err));
  // Error: 404
  ```

- 비동기 처리 시에 발생한 에러는 Promise 객체의 후속처리 메서드 catch를 사용해서도 처리할 수 있다.

- catch 메서드를 호출하면 내부적으로 then 메서드를 호출한다.

  ``` javascript
  const wrongUrl = 'https://jsonplaceholder.typicode.com/XXX/1';
  
  // 부적절한 url이 지정되어 에러발생
  promiseGet(wrongUrl)
  	.then(res => console.log(res))
  	.catch(err => console.log(err)); // Error: 404
  ```

  위의 예제는 내부적으로 아래와 같이 처리됨

  ``` javascript
  const wrongUrl = 'https://jsonplaceholder.typicode.com/XXX/1';
  
  promiseGet(wrongUrl)
  	.then(res => console.log(res))
  	.then(undefined, err => console.error(err));
  // Error: 404
  ```

- catch 메서드는 에러를 처리한다는 점에서 then 메서드의 두번째 콜백 함수와 동일하지만 차이가 있다.

  then 메서드의 두번째 콜백 함수는 비동기 처리에서 발생한 에러(reject함수가 호출된 상태)만을 캐치한다. => then 메서드 내부의 에러캐치 불가능

  ``` javascript
  promiseGet('https://jsonplaceholder.typicode.com/todos/1')
  	.then(res => console.x(res), err => console.error(err));
  // 두번째 콜백함수는 첫번째 콜백함수에서 발생한 에러를 캐치하지 못 한다.
  ```

- catch 메서드는 비동기 처리에서 발생한 에러(reject함수 호출상태)뿐 아니라 then 메서드 내부에서 발생한 에러도 캐치한다.

  ``` javascript
  promiseGet('https://jsonplaceholder.typicode.com/todos/1')
  	.then(res => console.x(res));
  	.catch(e => console.error(e));
  // TypeError : console.x is not a function
  ```

- then 메서드에 두번째 콜백함수를 전달하는 것보다 catch메서드를 사용하면 가독성이 더 좋다.

</br></br>

## 프로미스 체이닝

- 비동기 함수의 처리 결과를 가지고 다른 비동기 함수를 호출해야 하는 경우

  비동기 함수의 호출이 중첩(nesting)이 되어 복잡도가 높아지는 콜백 헬이 발생

- Promise 객체를 반환한 비동기 함수는 프로미스 후속 처리 메서드인 then, catch, finally 메서드를 사용할 수 있다. 

  > 후속 처리 메서드는 모두 Promise 객체를 반환한다. 

- 후속 처리 메서드를 체이닝(chaining)하여 호출할 수 있다. 

  이로써 콜백 헬을 해결

  ``` javascript
  const url = 'https://jsonplaceholder.typicode.com';
  
  // id가 1인 post의 userId를 취득
  promiseGet(`${url}/posts/1`)
  	//취득한 post의 userId로 user 정보를 취득
  	.then(({ userId }) => promiseGet(`${url}/users/${userId}`))
  	.then(userInfo => console.log(userInfo))
  	.catch(err => console.error(err));
  ```

</br></br>

## 프로미스의 정적 메서드

- Promise는 주로 생성자 함수로 사용되지만 함수도 객체이므로 메서드를 가질 수 있다. Promise 객체의 5가지 정적 메서드?
  - Promise.resolve / Promise.reject
  - Promise.all
  - Promise.race
  - Promise.allSettled

</br>

### Promise.resolve / Promise.reject

---

- Promise.resolve 와 Promise.reject 메서드는 

  이미 존재하는 값을 래핑하여 Promise 객체를 생성하기 위해 사용한다.

- 정적 메서드 Promise.resolve 메서드는 

  인자로 전달된 값을 resolve하는 Promise 객체를 생성한다.

  ``` javascript
  // 배열을 resolve 하는 Promise 객체를 생성
  const resolvedPromise = Promise.resolve([1, 2, 3]);
  resolvedPromise.then(console.log); // [1, 2, 3]
  ```

  아래와 동일

  ```javascript
  const resolvedPromise = new Promise(resolve => resolve([1, 2, 3]));
  resolvedPromise.then(console.log); // [1, 2, 3]
  ```

- Promise.reject 메서드는 인자로 전달된 값을 reject하는 Promise 객체를 생성한다.

  ``` javascript
  // 에러 객체를 reject하는 Promise 객체를 생성
  const rejectedPromise = Promise.reject(new Error('Error!'));
  rejectedPromise.catch(console.log); // Error: Error!
  ```

  아래와 동일

  ```javascript
  const rejectedPromise = new Promise((_, reject) => reject(new Error('Error!')));
  rejectedPromise.catch(console.log); // Error: Error!
  ```

</br>

### Promise.all

---

- Promise.all 메서드는 Promise 객체를 요소로 갖는 배열 등의 이터러블을 인자로 전달받는다.

- 전달받은 모든 Promise 객체를 모두 연속적으로 처리하고 

  처리 결과를 resolve하는 새로운 프로미스를 반환한다.

  ``` javascript
  Promise.all([
      new Promise(resolve => setTimeout(() => resolve(1), 3000)), // 1
      new Promise(resolve => setTimeout(() => resolve(2),2000)), //2
      new Promise(resolve => setTimeout(() => resolve(3), 1000)) //3
  ]).then(console.log) // [1, 2, 3]
    .catch(console.error);
  ```

  - 위의 Promise.all 메서드는 3개의 Promise 객체를 요소로 갖는 배열을 받는다. 각각의 Promise 객체의 동작은
    - 첫번째 Promise 객체는 3초 후에 1을 resolve하여 처리 결과를 반환한다.
    - 두번째 Promise 객체는 2초 후에 2을 resolve하여 처리 결과를 반환한다.
    - 세번째 Promise 객체는 1초 후에 3을 resolve하여 처리 결과를 반환한다.

- Promise.all 메서드는 **전달받은 모든 Promise 객체를 연속적으로 처리**한다.

- Promise.all은 배열 내 모든 Promise 객체의 resolve 또는 첫번째 reject를 기다린다.

- 모든 Promise 객체의 처리가 성공하면 모든 Promise 객체가 resolve한 처리 결과를 배열에 담아 resolve하는 새로운 Promise 객체를 반환한다.

  > Promise.all메서드가 반환하는 Promise 객체는 나중에 처리되는 여부와 상관없이 차례대로 처리순서가 보장된다.

- Promise 객체의 처리가 하나라도 실패하면 가장 먼저 실패한 Promise 객체가 reject한 에러를 reject하는 새로운 Promise 객체를 즉시 반환한다.

  ``` javascript
  Promise.all([
      new Promise((_, reject) => setTimeout(() => reject(new Error('Error 1')), 3000)),
    new Promise((_, reject) => setTimeout(() => reject(new Error('Error 2')), 2000)),
    new Promise((_, reject) => setTimeout(() => reject(new Error('Error 3')), 1000))
  ]).then(console.log)
    .catch(console.log); // Error: Error 3
  // 세번째 Promise 객체가 가장 먼저 실패하므로 세번재 Promise 객체가 reject한 에러가 catch 메서드로 전달된다.
  ```

- Promise.all 메서드는 인수로 전달받은 이터러블의 요소가 Promise 객체가 아닌 경우 Promise.resolve 메서드를 통해 Promise 객체로 래핑된다.

  ``` javascript
  Promise.all([
      1,  // => Promise.resolve(1)
      2,  // => Promise.resolve(2)
      3  // => Promise.resolve(3)
  ]).then(console.log) // [1, 2, 3]
    .catch(console.log);
  ```

- 아래의 예제의 Promise.all 메서드는 PromiseGet 함수가 반환한 3개의 Promise 객체의 배열을 인수로 전달받고

  이 Promise 객체들을 모두 연속적으로 처리한다.

  모든 Promise 객체의 처리가 성공하면 Promise.all  메서드는 

  각 Promise 객체가 resolve한 3개의 사용자 객체가 담긴 배열을 resolve하는 새로운 Promise객체를 반환하고 

  후속 처리 메서드 then에는 3개의 사용자 객체가 담긴 배열이 전달된다.

  ``` javascript
  // GET 요청을 위한 비동기 함수
  const promiseGet = url => {
      return new Promise((resolve, reject) => {
          const xhr = new XMLHttpRequest();
          xhr.open('GET',url);
          xhr.send();
          
          xhr.onload = () => {
              if (xhr.status === 200) {
                  // 성공적으로 응답 전달받으면 resolve 함수 호출
                  resolve(JSON.parse(xhr.response));
              } else {
                  // 에러 처리를 위해 reject 함수 호출
                  reject(new Error(xhr.status));
              }
          };
      });
  };
  
  const githubIds = ['jeresig', 'ahejlsberg', 'ungmo2'];
  
  Promise.all(githubIds.map(id => promiseGet(`https://api.github.com/users/${id}`)))
  // [user, user, user] => Promise
      .then(users => users.map(user => users.name));
  // ["John Resing", "Anders Hejiburg", "Ungmo Lee"]
  	.then(console.log)
      .catch(console.error);
  ```

</br>

### Promise.race

---

- Promise.race 메서드는 Promise.all 메서드와 동일하게 Promise 객체를 요소로 갖는 배열등의 이터러블을 인자로 전달받는다.

- Promise.race 메서드는 Promise.all 메서드처럼 모든 Promise 객체를 연속적으로 처리하는 것이 아니라 

  **가장 먼저 처리된 Promise 객체가 resolve한 처리 결과**를 resolve하는 새로운 Promise 객체를 반환한다. 

  ``` javascript
  Promise.race([
      new Promise(resolve => setTimeout(() => resolve(1), 3000)),  // 1
      new Promise(resolve => setTimeout(() => resolve(2), 2000)), //2
      new Promise(resolve => setTimeout(() => resolve(3), 1000)) //3
      
  ]).then(console.log) // 3
     .catch(console.log)
  ```

- 에러가 발생하는 경우, Promise.all 메서드와 동일하게 처리된다.

  > Promise.race 메서드에 전달된 Promise 객체의 처리가 하나라도 실패하면 가장 먼저 실패한 Promise 객체가 reject한 에러를 reject하는 새로운 Promise 객체를 반환.

  ``` javascript
  Promise.race([
      new Promise((_, reject) => setTimeout(new Error('Error 1'), 3000)),
      new Promise((_, reject) => setTimeout(new Error('Error 2'), 2000)),
      new Promise((_, reject) => setTimeout(new Error('Error 3'), 1000))
  ]).then(console.log).catch(console.log); // Error: Error 3
  ```

  </br>

### Promise.allSettled

---

- Promise.allSettled 메서드는 Promise 객체를 요소로 갖는 배열 등의 이터러블을 인자로 전달받는다.

- 전달받은 모든 Promise 객체를 모두 연속적으로 처리하고 그 **처리 결과를 배열**로 반환한다.

  ``` javascript
  Promise.allSettled([
      new Promise(resolve => setTimeout(() => resolve(1), 2000)),
      new Promise((_, reject) => setTimeout(() => reject(new Error('Error!')), 1000))
  ]).then(console.log);
  
  /*
  [
  	{status: "fulfilled", value: 1},
  	{status: "rejected", reason: Error: Error! at <anonymous>:3:60}
  ]
  */
  ```

- Promise.allSettled 메서드가 반환한 배열에는 

  fulfilled 또는 rejected 상태와는 상관없이 

  Promise.allSettled 메서드가 **인수로 전달받은 모든 Promise 객체들의 처리 결과**가 담겨있다.

- 모든 Promise 객체의 처리 결과를 나타내는 객체는

   Promise 객체가 **fulfilled 상태**인 경우,

  처리상태를 나타내는 **status 프로퍼티**와 처리 결과를 나타내는 **value 프로퍼티**를 갖는다.

- Promise 객체가 **rejected 상태**인 경우,

   처리 상태를 나타내는 **status 프로퍼티**와 에러를 나타내는 **reason 프로퍼티**를 갖는다.

  ``` javascript
  [
      // 프로미스가 fulfilled 상태인 경우, 프로미스의 처리 결과
      { status: "fulfilled", value: 1 },
      // 프로미스가 rejected 상태인 경우, 프로미스의 처리 결과
      { status: "rejected", reason: Error: Error! at <anonymous>:3:60}
  ]
  ```

  </br></br>

## fetch

- fetch 함수는 XMLHttpRequest 객체와 마찬가지로 **HTTP 요청 전송 기능을 제공하는 Web API** 이다.

- fetch 함수는 XMLHttpRequest 객체보다 사용법이 간단하고 **프로미스를 지원**하기 때문에 비동기 처리를 위한 콜백 패턴의 단점에서 자유롭다.

- fetch 함수는 비교적 최근에 추가된 Web API로서 인터넷 익스플로어를 제외한 대부분의 브라우저에서 제공하고 있다.

- fetch 함수에는 HTTP 요청을 전송할 URL과 HTTP 요청 메서드, HTTP 요청 헤더, 페이로드 등을 설정한 객체를 전달한다.

  ``` javascript
  const promise = fetch(url[, options])
  ```

- fetch 함수는 HTTP 응답을 나타내는 Response 객체를 래핑한 Promise 객체를 반환한다.

- fetch 함수에 첫번째 인수로 HTTP 요청을 할 URL만 전달하면 GET 요청을 전송한다.

  ```javascript
  fetch('https://jsonplaceholder.typicode.com/todos/1')
  	.then(response => console.log(response));
  ```

- fetch 함수는 HTTP 응답을 나타내는 Response 객체를 래핑한 프로미스를 반환하므로 후속처리 메서드 then을 통해 프로미스가 resolve한 Response 객체를 전달 받을 수 있다. 
- Response 객체는 HTTP응답을 나타내는 다양한 프로퍼티를 제공한다.

- Reponse.prototype 에는 Response 객체에 포함되어 있는 HTTP 응답 몸체(body)를 위한 다양한 메서드를 제공한다.

  > fetch 함수가 반환한 프로미스가 래핑하고 있는 HTTP 응답 몸체를 취득하려면,
  >
  > Response.prototype.json 메서드를 사용한다.
  >
  > Reponse.prototype.json 메서드는 Response 객체에서 HTTP 응답몸체(response.body)를 취득하여 역직렬화한다.

  ```javascript
  fetch('https://jsonplaceholder.typicode.com/todos/1')
  	// response는 HTTP 응답을 나타내는 Response 객체이다.
  	// json 메서드를 사용하여 Response 객체에서 HTTP 응답 몸체를 취득하여 역직렬화 한다.
  	.then(response => response.json())
  	// json은 역직렬화된 HTTP 응답몸체이다.
  	.then(json => console.log(json));
  	// {userId: 1, id: 1, title: "..", completed: false}
  ```

  > fetch 함수를 통해 POST/PATCH/DELETE 요청을 전송해보자.
  >
  > fetch 함수에 첫번째 인수로 HTTP 요청을 전송할 URL과 두번째 인수로 HTTP 요청 메서드, HTTP 요청 헤더, 페이로드 등을 설정한 객체를 전달한다.

  ``` javascript
  const request = {
    get(url) {
       return fetch(url);
    },
    post(url, payload) {
        return fetch(url, {
            method: 'POST',
            headers: { 'content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });
    },
    patch(url, payload) {
        return fetch(url, {
            method: 'PATCH',
            headers: { 'content-Type': 'application/json'},
            body: JSON.stringify(payload)
        });
    },
    delete(url) {
        return fetch(url, { method: 'DELETE' });
    }
  };
  
  request.get('https://jsonplaceholder.typicode.com/todos/1')
  	.then(response => response.json())
  	.then(todos => console.log(todos))
  	.catch(err => console.error(err));
  // {userId: 1, id: 1, title: 'text', completed: false}
  
  request.post('https://jsonplaceholder.typicode.com/todos',
  {
      userId: 1,
      title: 'JavaScript',
      completed: false
  })
  	.then(reponse => response.json())
  	.then(todos => console.log(todos))
  	.catch(err => console.err(err))
  // {userId: 1, title: "JavaScript", completed: false, id: 201}
  
  request.patch('https://jsonplaceholder.typicode.com/todos/1', {
    completed: true
  })
    .then(response => response.json())
    .then(todos => console.log(todos))
    .catch(err => console.error(err));
  // {userId: 1, id: 1, title: "delectus aut autem", completed: true}
  
  request.delete('https://jsonplaceholder.typicode.com/todos/1')
    .then(response => response.json())
    .then(todos => console.log(todos))
    .catch(err => console.error(err));
  // {}
  ```

  

